## Task and Models

Our task is to focus on activity recognition, at least for now. We will create a few models to get a baseline

#### Random Forest
A random forest may be able to accurately predict activities, as this data is somewhat complex and noisy

#### KNN
K-Nearest Neighbors is likely not a good choice due to the curse of dimensionality

#### Logistic/Linear Regression
The relationships here aren't linear, so this likely wouldn't be a good choice

#### Decision Trees
These may be a good option due to the breadth of the data, but the data set is very large which may cause issues

#### DNN
Deep neural networks handle high-dimensional data well, but do not take into account sequencing. Due to this, DNNs are not a good choice

#### CNN
Convolutional neural networks are good for grid-like structures (images) and for spotting patterns in certain sectors of data. This may be a possible option, but likely wouldn't perform well. It's main use cases involve image, video, or natural language processing.

#### LSTM
Long Short-Tern Memory models are good for understanding sequential data, context, and long term dependencies. This is likely the best choice for the job (out of the deep models)


In [7]:
import src.utils.create_models_util as models_util

# LSTM

In [3]:
FILE_PATH = "../processed_data/SDHAR/final_processed_data_ALL_DAYS.csv"
SAVE_PATH = '../models/SDHAR/LSTM_first_iteration.keras'
TARGET_COLUMN = 'activity_user_1'  # We will predict the activity for User 1
WINDOW_SIZE = 60  # How many past time steps to look at (e.g., 60 * 2s = 120 seconds of history)
STEP_SIZE = 30    # How far to slide the window forward each time

models_util.train_and_save_lstm(FILE_PATH, SAVE_PATH, TARGET_COLUMN, WINDOW_SIZE, STEP_SIZE)

Loading final processed data...
Separating data...
Creating sliding windows (size=60, step=30)...
  - Windowed X shape: (67467, 60, 41)
  - Windowed y shape: (67467, 18)
Splitting data into training and test sets...
  - Training set size: 53973
  - Test set size: 13494
Building the LSTM model...


C:\Users\jesse\AppData\Local\Programs\Python\Python313\Lib\site-packages\keras\src\layers\rnn\rnn.py:199: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ lstm (LSTM)                     │ (None, 60, 64)         │        27,136 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout (Dropout)               │ (None, 60, 64)         │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ lstm_1 (LSTM)                   │ (None, 64)             │        33,024 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout_1 (Dropout)             │ (None, 64)             │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense (Dense)                   │ (None, 18)             │         1,170 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 61,330 (239.57 KB)

 Trainable params: 61,330 (239.57 KB)

 Non-trainable params: 0 (0.00 B)


Training the model...
Epoch 1/10
380/380 ━━━━━━━━━━━━━━━━━━━━ 56s 141ms/step - accuracy: 0.7179 - loss: 1.0096 - val_accuracy: 0.8025 - val_loss: 0.6839
Epoch 2/10
380/380 ━━━━━━━━━━━━━━━━━━━━ 56s 148ms/step - accuracy: 0.8182 - loss: 0.6165 - val_accuracy: 0.8486 - val_loss: 0.5262
Epoch 3/10
380/380 ━━━━━━━━━━━━━━━━━━━━ 59s 155ms/step - accuracy: 0.8424 - loss: 0.5152 - val_accuracy: 0.8529 - val_loss: 0.5393
Epoch 4/10
380/380 ━━━━━━━━━━━━━━━━━━━━ 62s 164ms/step - accuracy: 0.8688 - loss: 0.4291 - val_accuracy: 0.8785 - val_loss: 0.3889
Epoch 5/10
380/380 ━━━━━━━━━━━━━━━━━━━━ 55s 144ms/step - accuracy: 0.8788 - loss: 0.4005 - val_accuracy: 0.8918 - val_loss: 0.3478
Epoch 6/10
380/380 ━━━━━━━━━━━━━━━━━━━━ 39s 102ms/step - accuracy: 0.8932 - loss: 0.3516 - val_accuracy: 0.8961 - val_loss: 0.3330
Epoch 7/10
380/380 ━━━━━━━━━━━━━━━━━━━━ 37s 98ms/step - accuracy: 0.8948 - loss: 0.3467 - val_accuracy: 0.9040 - val_loss: 0.3008
Epoch 8/10
380/380 ━━━━━━━━━━━━━━━━━━━━ 42s 111ms/step - accu

C:\Users\jesse\AppData\Local\Programs\Python\Python313\Lib\site-packages\sklearn\metrics\_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
C:\Users\jesse\AppData\Local\Programs\Python\Python313\Lib\site-packages\sklearn\metrics\_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
C:\Users\jesse\AppData\Local\Programs\Python\Python313\Lib\site-packages\sklearn\metrics\_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, mo

# Random Forest

In [11]:
FILE_PATH = "../processed_data/SDHAR/final_processed_data_ALL_DAYS.csv"
SAVE_PATH = "../models/SDHAR/RandomForest_first_iteration.joblib"
TARGET_COLUMN = 'activity_user_1'
WINDOW_SIZE = 60  # 120 seconds of history
STEP_SIZE = 30

models_util.train_and_save_rf(FILE_PATH, SAVE_PATH, TARGET_COLUMN, WINDOW_SIZE, STEP_SIZE)

Loading final processed data...
Separating Features and Target...
Creating sliding windows (size=60, step=30)...
  - Initial windowed X shape: (67467, 60, 41)
Flattening window data...
  - Flattened X shape: (67467, 2460)
Splitting data into training and test sets...
  - Training set size: 53973
  - Test set size: 13494

Building and training the Random Forest model...

Evaluating the model on the test set...

Test Accuracy: 97.90%

Classification Report:
                   precision    recall  f1-score   support

BATHROOM ACTIVITY       0.94      0.92      0.93       381
           CHORES       0.99      0.88      0.93        83
             COOK       0.96      0.89      0.93       139
      DISHWASHING       1.00      0.88      0.93        16
            DRESS       0.75      0.36      0.49        33
              EAT       0.92      0.98      0.95       609
          LAUNDRY       0.00      0.00      0.00         1
 MAKE SIMPLE FOOD       0.93      0.72      0.81       104
        

C:\Users\jesse\AppData\Local\Programs\Python\Python313\Lib\site-packages\sklearn\metrics\_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
C:\Users\jesse\AppData\Local\Programs\Python\Python313\Lib\site-packages\sklearn\metrics\_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
C:\Users\jesse\AppData\Local\Programs\Python\Python313\Lib\site-packages\sklearn\metrics\_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, mo

# Decision Tree

In [4]:
FILE_PATH = "../processed_data/SDHAR/final_processed_data_ALL_DAYS.csv"
SAVE_PATH = "../models/SDHAR/DecisionTree_first_iteration.joblib"
TARGET_COLUMN = 'activity_user_1'
WINDOW_SIZE = 60
STEP_SIZE = 30

models_util.train_and_save_dt(FILE_PATH, SAVE_PATH, TARGET_COLUMN, WINDOW_SIZE, STEP_SIZE)

Loading final processed data...
Separating Features and Target...
Creating sliding windows (size=60, step=30)...
  - Initial windowed X shape: (67467, 60, 41)
Flattening window data...
  - Flattened X shape: (67467, 2460)
Splitting data into training and test sets...
  - Training set size: 53973
  - Test set size: 13494
Building and training the Decision Tree model...
  - Model training complete!
Evaluating the model on the test set...

Test Accuracy: 95.91%

Classification Report:
                   precision    recall  f1-score   support

BATHROOM ACTIVITY       0.90      0.89      0.89       381
           CHORES       0.81      0.78      0.80        83
             COOK       0.81      0.78      0.80       139
      DISHWASHING       0.82      0.56      0.67        16
            DRESS       0.23      0.24      0.24        33
              EAT       0.88      0.90      0.89       609
          LAUNDRY       0.00      0.00      0.00         1
 MAKE SIMPLE FOOD       0.70      0.64  

# Model analysis

### LSTM
The LSTM model has disappointing performance so far. While it has an overall accuracy of 92%, it completely fails to identify less frequent activities. There are some key improvements that can be made here, however. First, and likely most importantly, we can rebalance the activity distribution. Using techniques like SMOTE, we can "add" more of these activities to better train the model. We can also experiment with tuning the models hyperparameters and/or number of epochs. This should allow us to increase performance with a deep model

### Random Forest and Decision Tree
Both of these models performed similar. As we can see,  they had accuracies of 97.9% and 95.9% respectively. However, these numbers don't tell the whole truth. Because longer, more predictable activities like SLEEP, OUT HOME, and WATCH TV dominate the dataset, most of this performance comes from this. We can see that the models perform poorly on shorter, infrequent, less predicable activities like PET, TAKE MEDS, and DRESS. Random Forest performs definitively better than the Decision Tree. This was somewhat expected, and for the future, we will likely focus on using Random Forests for shallow models. Similarly to the LSTM model, the best thing we can do to increase performance is utilizing SMOTE to oversample minority activities. This should make a big difference in predicting these activities
